R/make-traceability-matrix.R

```diff
@@ -376,7 +391,7 @@ filter_symbol_functions <- function(funcs){
 get_toplevel_assignments <- function(pkg_source_path){
   r_files <- list.files(
     file.path(pkg_source_path, "R"),
-    full.names = TRUE, pattern = "\\.[Rr]$", recursive = TRUE
+    full.names = TRUE, pattern = "\\.(?i)[rsq]$", recursive = TRUE
```

Comment:
Makes sense. Thanks for the background in <https://github.com/metrumresearchgroup/mpn.scorecard/issues/46#issuecomment-1724251388> and 380638e.

---

Poking around a bit in the R code base: `install` [calls](https://github.com/wch/r-source/blob/546e5aaa701f711c0afa64e091e3c8f45c7ba85d/src/library/tools/R/admin.R#L249-L248) a helper, [`list_files_with_type`](https://github.com/wch/r-source/blob/546e5aaa701f711c0afa64e091e3c8f45c7ba85d/src/library/tools/R/utils.R#L151-L188), to get the code files. That handles the different extensions (via [`.make_file_exts`](https://github.com/wch/r-source/blob/546e5aaa701f711c0afa64e091e3c8f45c7ba85d/src/library/tools/R/utils.R#L1706-L1727)), and also handles the [OS-specific subdirectory case](https://github.com/metrumresearchgroup/mpn.scorecard/pull/42#discussion_r1326392508).

`list_files_with_type` is exported.

```R
> tools::list_files_with_type("R", "code")
 [1] "R/aaa.R"                      "R/add-pkg-info.R"
 [3] "R/calc-overall-scores.R"      "R/check.R"
 [5] "R/coverage.R"                 "R/create-extra-notes.R"
 [7] "R/format-report.R"            "R/make-traceability-matrix.R"
 [9] "R/metadata.R"                 "R/render-scorecard-summary.R"
[11] "R/render-scorecard.R"         "R/riskmetric.R"
[13] "R/score-pkg.R"                "R/scorecard.R"
[15] "R/util.R"                     "R/unix/foo.R"
```

I think it'd be a good idea to switch to using `list_files_with_type`. That would align our subdirectory handling R's actual behavior: only list files in a OS-matching subdirectory. And in general using the same code as R keeps our logic from getting out of sync with R's (even if unlikely).

Reply:
I like this change. However I have one concern:

In the usage part of the help doc we see:
```r
list_files_with_type(dir, type, all.files = FALSE,
                     full.names = TRUE, OS_subdirs = .OStype())
```

where on Metworx..
```r
> tools:::.OStype()
[1] "unix"
```

I actually think we want all subdirectories to be checked here, since the main purpose is for the traceability matrix:
```r
tools::list_files_with_type("R", "code", OS_subdirs = c("unix", "linux","windows"))
```

Reply:
> Im not sure about this though. It could lead to duplicate exports, so perhaps only the actual operating system is what we're looking for.

Right, in my view the advantage of this is that we match what R itself actually loads, ignoring the files in the subdirectories for other systems.  That solves the "multiple functions" issues that you mention above (and mentioned earlier at <https://github.com/metrumresearchgroup/mpn.scorecard/pull/42#discussion_r1326461354>).  OS-specific implementations for the same functions will be the typical case for these directories.

So yes, I think this call should be restricted to the current system (i.e. the default).